File: 4.controls/mobile/Engine/Source/Objects/CCTile3DButton.cpp
/*-----------------------------------------------------------
 * 2c - Cross Platform 3D Application Framework
 *-----------------------------------------------------------
 * Copyright © 2010 – 2011 France Telecom
 * This software is distributed under the Apache 2.0 license,
 * see the "license.txt" file for more details.
 *-----------------------------------------------------------
 * File Name   : CCTile3DButton.cpp
 *-----------------------------------------------------------
 */

#include "CCDefines.h"
#include "CCObjects.h"
#include "CCTextureFontPage.h"


CCTile3DButton::CCTile3DButton(CCSceneBase *scene)
{
    construct( scene );
}


CCTile3DButton::CCTile3DButton(CCSceneBase *scene, const float width, const float height, const char *text)
{
    construct( scene );
    
    setupBase( width, height );
    
    if( text )
    {
        textModel->setText( text, height );
    }
}


CCTile3DButton::CCTile3DButton(CCSceneBase *scene, const float width, const char *textureName)
{
    construct( scene );
    
    const CCTextureBase *texture = gEngine->textureManager->getTexture( textureName, Resource_Packaged );
    const float aspectRatio = texture->getImageWidth() / texture->getImageHeight();
    const float height = width / aspectRatio;
    setupBase( width, height );

    setBaseTexture( textureName, Resource_Packaged );
}


CCTile3DButton::CCTile3DButton(CCSceneBase *scene, const char *text, const float height, const bool centered)
{
    construct( scene );
    
    textModel->setText( text, height );
    
    const float width = textModel->getWidth();
    
    setHCollisionBounds( width * 0.5f, height * 0.5f, CC_SMALLFLOAT );
    
    textModel->setCentered( centered );
    
    if( centered == false )
    {
        translate( collisionBounds.x, 0.0f, 0.0f );
    }
}


void CCTile3DButton::construct(CCSceneBase *scene)
{
    AddFlag( collideableType, collision_ui );

    if( scene != NULL )
    {
        setScene( scene );
    }
    
    renderDepth = false;
    setTransparent();
    
    model = new CCModelBase();
    
    baseModel = new CCModelBase();
    model->addModel( baseModel );
    setBaseColour( CCColour() );
    colourInterpolator.setDuration( 0.5f );
    baseSquare = NULL;
    
    textModel = new CCModelText( this );
    
    allowTouchRotation( false );
    touchRotationMagnitude = 0.0f;
    touchRotationSpeed = 1.0f;
    
    if( scale == NULL )
    {
        CCVector3FillPtr( &scale, 0.0f, 0.0f, 1.0f );
    }
    scaleInterpolator.setup( scale, 1.0f );

    allowTouchMovement( false );

    touchDepressInterpolator.setDuration( 0.125f );
    setTouchDepressRange( 1.0f );
    touchDepressDepth = 3.0f;

    CCUpdateCollisions( this );
}


void CCTile3DButton::destruct()
{
    onPress.deleteObjectsAndList();
    onRelease.deleteObjectsAndList();
    onLoss.deleteObjectsAndList();

    super::destruct();
}

// CCRenderable
void CCTile3DButton::refreshModelMatrix()
{
	if( updateModelMatrix )
	{
        CCMatrixLoadIdentity( modelMatrix );	
        CCMatrixTranslate( modelMatrix, position.x, position.y, position.z );
        CCMatrixTranslate( modelMatrix, touchDepressPosition.x, touchDepressPosition.y, touchDepressPosition.z );
        
		if( scale != NULL )
		{	
			CCMatrixScale( modelMatrix, scale->x, scale->y, scale->z );
		}

        if( rotation.x != 0.0f )
        {
            CCMatrixRotate( modelMatrix, rotation.x, 1.0f, 0.0f, 0.0f );
        }

        if( rotation.y != 0.0f )
        {
            CCMatrixRotate( modelMatrix, rotation.y, 0.0f, 1.0f, 0.0f );
        }
        
		if( touchRotationMagnitude != 0.0f )
        {
            CCMatrixRotate( modelMatrix, touchRotationMagnitude * 20.0f, 1.0f, 0.0f, 0.0f );
        }
        
		updateModelMatrix = false;
		dirtyWorldMatrix();
	}
}


// CCSceneObject
void CCTile3DButton::update(const CCTime &gameTime)
{
    super::update( gameTime );

    if( scale != NULL )
    {
        const float speed = gameTime.delta * 1.0f;
        if( scaleInterpolator.update( speed ) )
        {
            dirtyModelMatrix();
        }
        else
        {
            DELETE_POINTER( scale );
        }
    }

    // Touch depress: On update
    if( touchDepressRange > 0.0f )
    {
        if( touchDepressInterpolator.update( gameTime.delta ) )
        {
            dirtyModelMatrix();
        }
    }

    if( touchRotationAllowed )
    {
        const float speed = touching || touchReleased ? 3.0f : touchRotationSpeed;
        if( touchRotationInterpolator.update( gameTime.delta * speed ) )
        {
            const float magnitudeSquared = rotation.x * rotation.x + rotation.y * rotation.y;
            touchRotationMagnitude = sqrtf( magnitudeSquared );
            touchRotationMagnitude = MIN( touchRotationMagnitude, 1.0f );

            dirtyModelMatrix();
        }
    }

    if( touching )
    {
        touchingTime += gameTime.real;
    }
    else if( touchReleased )
    {
        if( touchDepressInterpolator.finished() && touchRotationInterpolator.finished() )
        {
            handleTouchRelease();
        }
    }

    colourInterpolator.update( gameTime.delta );

    if( textModel != NULL )
    {
        textModel->colourInterpolator.update( gameTime.delta );
    }
}


void CCTile3DButton::renderModels(const bool alpha)
{
    if( alpha )
    {
        if( renderDepth )
        {
            glEnable( GL_DEPTH_TEST );
            super::renderModels( alpha );
            glDisable( GL_DEPTH_TEST );
        }
        else
        {
            super::renderModels( alpha );
        }
    }
    else
    {
        if( renderDepth )
        {
            super::renderModels( alpha );
        }
        else
        {
            glDisable( GL_DEPTH_TEST );
            super::renderModels( alpha );
            glEnable( GL_DEPTH_TEST );
            
        }
    }
}


void CCTile3DButton::setupBase(const float width, const float height)
{
    if( baseSquare == NULL )
    {
        baseSquare = new CCPrimitiveSquare();
        baseModel->addPrimitive( baseSquare );
    }

    const float hWidth = width * 0.5f;
    const float hHeight = height * 0.5f;
    setHCollisionBounds( hWidth, hHeight, CC_SMALLFLOAT );
    baseSquare->setupZFacing( collisionBounds.x, collisionBounds.y );
    CCUpdateCollisions( this );
}


void CCTile3DButton::setBaseTexture(const char *name, const CCResourceType resourceType)
{
    baseSquare->setTexture( name, resourceType );
}


void CCTile3DButton::setBaseColour(const CCColour &inColour)
{
    baseModel->setColour( inColour );
    colourInterpolator.setup( baseModel->getColour(), *baseModel->getColour() );
}


void CCTile3DButton::setBaseColourAlpha(const float inAlpha)
{
    baseModel->setColourAlpha( inAlpha );
    colourInterpolator.setup( baseModel->getColour(), *baseModel->getColour() );
}


const uint CCTile3DButton::handleProjectedTouch(const CCCameraProjectionResults &cameraProjectionResults,
                                                const CCSceneCollideable *hitObject, 
                                                const CCVector3 &hitPosition,
                                                const CCScreenTouches &touch, 
                                                const CCTouchAction touchAction)
{
    if( enabled == false )
    {
        return false;
    }

    if( hitObject == this && 
        ( touchAction == touch_pressed || ( touchMovementAllowed && CCControls::touchActionMoving( touchAction ) ) ) )
    {
        if( touching == false )
        {
            touching = true;
            touchingTime = 0.0f;
            onTouchPress();
        }
    }
    
    if( touching )
    {
        const float maxTimeHeld = 0.125f;
        
        if( touchAction == touch_pressed )
        {
            if( touch.timeHeld >= maxTimeHeld )
            {
                CCVector3 relativeHitPosition;
                relativeHitPosition.x = hitPosition.x - position.x;
                relativeHitPosition.y = hitPosition.y - position.y;
                float x = relativeHitPosition.x / collisionBounds.x;
                float y = relativeHitPosition.y / collisionBounds.y;
                touchActionPressed( x, y, touchAction );
            }
        }
        else if( touchMovementAllowed && CCControls::touchActionMoving( touchAction ) )
        {
            if( hitObject != this )
            {
                return handleProjectedTouch( cameraProjectionResults, hitObject, hitPosition, touch, touch_lost );
            }
            else
            {
                CCVector3 relativeHitPosition;
                relativeHitPosition.x = hitPosition.x - position.x;
                relativeHitPosition.y = hitPosition.y - position.y;
                float x = relativeHitPosition.x / collisionBounds.x;
                float y = relativeHitPosition.y / collisionBounds.y;
                touchActionPressed( x, y, touchAction );
            }
        }
        else
        {   
            // Ensure we have a good touch release
            if( touchAction == touch_released )
            {
                if( touchMovementAllowed == false )
                {
                    const float absDeltaX = fabsf( touch.totalDelta.x );
                    const float absDeltaY = fabsf( touch.totalDelta.y );
                    if( absDeltaX > CC_TOUCH_TO_MOVEMENT_THRESHOLD || absDeltaY > CC_TOUCH_TO_MOVEMENT_THRESHOLD )
                    {
                        return handleProjectedTouch( cameraProjectionResults, hitObject, hitPosition, touch, touch_lost );
                    }
                }
                
                // If we have a good first touch, register it.
                CCVector3 relativeHitPosition;
                relativeHitPosition.x = hitPosition.x - position.x;
                relativeHitPosition.y = hitPosition.y - position.y;
                float x = relativeHitPosition.x / collisionBounds.x;
                float y = relativeHitPosition.y / collisionBounds.y;
                touchActionPressed( x, y, touch_pressed );
            }

            touching = false;
            touchActionRelease( touchAction );
            
            if( touchAction == touch_released )
            {
                return 2;
            }
        }
    }
    
    return 0;
}


void CCTile3DButton::touchActionPressed(const float x, const float y, const CCTouchAction touchAction)
{
    if( touchRotationAllowed )
    {
        float clampedX = x;
        float clampedY = y;
        CCClampFloat( clampedX, -1.0f, 1.0f );
        CCClampFloat( clampedY, -1.0f, 1.0f );
        touchRotationInterpolator.pushV3( &rotation, CCVector3( -clampedY, clampedX, 0.0f ), true );
    }
    
    // Touch depress: On pressed
    if( touchDepressRange > 0.0f )
    {
        // Touch depress range specifies the area which the button is pressed down
        // If it's out of range the button is pushed back up
        if( fabsf( x ) < touchDepressRange && fabsf( y ) < touchDepressRange )
        {
            touchDepressInterpolator.pushV3( &touchDepressPosition, CCVector3( 0.0f, 0.0f, -touchDepressDepth ), true );
        }
        else
        {
            touchDepressInterpolator.pushV3( &touchDepressPosition, CCVector3(), true );
        }
    }

    if( touchAction > touch_pressed && touchAction < touch_released )
    {
        onTouchMove();
    }
}


void CCTile3DButton::touchActionRelease(const CCTouchAction touchAction)
{
    if( touchAction == touch_released )
    {
        touchReleased = true;
    }
    else 
    {
        handleTouchRelease();
        if( touchAction == touch_lost )
        {
            onTouchLoss();
        }
    }
}


void CCTile3DButton::handleTouchRelease()
{
    // If the touch has been released successfully we fire our callback
    if( touchReleased )
    {
        touchReleased = false;
        onTouchRelease();
    }

    if( touchRotationAllowed )
    {
        touchRotationInterpolator.pushV3( &rotation, CCVector3() );
    }

    // Touch depress: On releasse
    if( touchDepressRange > 0.0f )
    {
        touchDepressInterpolator.pushV3( &touchDepressPosition, CCVector3(), true );
    }
}

File: 4.controls/webgl/source/js/engine/objects/CCTile3DButton.js
/*-----------------------------------------------------------
 * 2c - Cross Platform 3D Application Framework
 *-----------------------------------------------------------
 * Copyright © 2011 - 2011 France Telecom
 * This software is distributed under the Apache 2.0 license.
 * http://www.apache.org/licenses/LICENSE-2.0.html
 *-----------------------------------------------------------
 * File Name   : CCTile3DButton.js
 * Description : A dynamic button widget.
 *
 * Created     : 09/08/11
 * Author(s)   : Ashraf Samy Hegab
 *-----------------------------------------------------------
 */

function CCTile3DButton(scene)
{
    this.construct( scene );
}
CopyPrototype( CCTile3DButton, CCTile3D, "CCTile3D" );


CCTile3DButton.prototype.construct = function(scene)
{
	this.CCTile3D_construct();
	
	if( scene )
	{
		this.setScene( scene );
	}
	
	this.renderDepth = false;
	this.setTransparent( true );

    var model = this.model = new CCModelBase();
    
    var baseModel = this.baseModel = new CCModelBase();
    model.addModel( baseModel );
	this.colourInterpolator = new CCInterpolatorColour();
	this.colourInterpolator.setDuration( 0.5 );
	this.setBaseColour( gColour.white() );
	this.baseSquare = false;
	
	this.textModel = new CCModelText( this );
	
	this.allowTouchRotation( false );
	this.touchRotationMagnitude = 0.0;
	this.touchRotationSpeed = 1.0;
	
	//this.scaleInterpolator.setup( scale, 1.0 );
	
	this.allowTouchMovement( false );

    this.touchDepressPosition = vec3.create();
    this.touchDepressInterpolator = new CCInterpolatorsV3();
    this.touchDepressInterpolator.speed = 3.0;
	this.setTouchDepressRange( 1.0 );
    this.touchDepressDepth = 2.0;
}


CCTile3DButton.prototype.setup = function(width, height, text)
{
	this.setupBase( width, height );
    
    if( text )
    {
        this.textModel.setText( text, height );
    }
	
	return this;
}

// Create tile with width and textures aspect ratio
CCTile3DButton.prototype.setupTexture = function(width, textureSrc)
{
	var TextureLoadCallback = function(tile, src) 
	{
		this.tile = tile;
		this.src = src;
	}
    TextureLoadCallback.prototype.callback = function(texture)
    {
        var aspectRatio = texture.width / texture.height;
		var height = width / aspectRatio;
		this.tile.setupBase( width, height );
		
		this.tile.setBaseTexture( this.src );
    }
	
	gEngine.textureManager.getTexture( textureSrc, new TextureLoadCallback( this, textureSrc ) );
	return this;
}

// Create tile with the text height
CCTile3DButton.prototype.setupText = function(text, height, centered)
{
	var textModel = this.textModel;
	textModel.setText( text, height );
    
    var width = textModel.getWidth();
    this.setHCollisionBounds( width * 0.5, height * 0.5, CC_SMALLFLOAT );
    
    textModel.setCentered( centered );
    
    if( centered == false )
    {
        this.translate( collisionBounds.x, 0.0, 0.0 );
    }
	
	return this;
}


CCTile3DButton.prototype.destruct = function(width, height)
{
}


// CCRenderable
CCTile3DButton.prototype.refreshModelMatrix = function()
{
	if( this.updateModelMatrix )
    {
        this.updateModelMatrix = false;
        
        var modelMatrix = this.modelMatrix;
        mat4.identity( modelMatrix );
		
        mat4.translate( modelMatrix, this.position );
        mat4.translate( modelMatrix, this.touchDepressPosition );

        mat4.scale( modelMatrix, this.scale );

        var rotation = this.rotation;
        mat4.rotate( modelMatrix, rotation[0], [1.0, 0.0, 0.0] );
        //mat4.rotate( modelMatrix, rotation[1], [0.0, 1.0, 0.0] );
        gTools.CCMatrixRotate( modelMatrix, rotation[1], 0.0, 1.0, 0.0 );
    }
}


CCTile3DButton.prototype.update = function(delta)
{
    if( this.touchDepressRange > 0.0 )
    {
        if( this.touchDepressInterpolator.update( delta ) )
        {
            this.updateModelMatrix = true;
        }
    }

    if( this.touching )
    {
        this.touchingTime += delta;
    }
    else if( this.touchReleased )
    {
        if( this.touchDepressInterpolator.finished() )
        {
            this.handleTouchRelease();
        }
    }
	
	this.colourInterpolator.update( delta );
	
	if( this.textModel != false )
    {
        this.textModel.colourInterpolator.update( delta );
    }
}


CCTile3DButton.prototype.setupBase = function(width, height)
{
    if( this.baseSquare == false )
    {
        var baseSquare = this.baseSquare = new CCPrimitiveSquare();
        this.baseModel.addPrimitive( baseSquare );
    }

    var hWidth = width * 0.5;
    var hHeight = height * 0.5;
    this.setHCollisionBounds( hWidth, hHeight, CC_SMALLFLOAT );
	this.baseSquare.setScale( width, height, 1.0 );
    
    this.refreshCollisions();
}


CCTile3DButton.prototype.setBaseTexture = function(src)
{
	this.baseSquare.setTexture( src );
}


CCTile3DButton.prototype.setBaseColour = function(inColour)
{
	this.baseModel.setColour( inColour );
	this.colourInterpolator.setup( this.baseModel.colour );
}


CCTile3DButton.prototype.handleProjectedTouch = function(hitObject, hitPosition, 
                                                 touch, touchAction)
{
    if( this.enabled == false )
    {
        return false;
    }

    if( hitObject == this &&
        ( touchAction == touch_pressed || ( this.touchMovementAllowed && CCControls.touchActionMoving( touchAction ) ) ) )
    {
        if( this.touching == false )
        {
            this.touching = true;
            this.touchingTime = 0.0;
            this.onTouchPress();
        }
    }
    
    if( this.touching )
    {
        var position = this.position;
        var collisionBounds = this.collisionBounds;
        var maxTimeHeld = 0.0;
        
        if( touchAction == touch_pressed )
        {
            if( touch.timeHeld >= maxTimeHeld )
            {
                var relativeHitPosition = vec3.create();
                relativeHitPosition[0] = hitPosition[0] - position[0];
                relativeHitPosition[1] = hitPosition[1] - position[1];
                var x = relativeHitPosition[0] / collisionBounds[0];
                var y = relativeHitPosition[1] / collisionBounds[1];
                this.touchActionPressed( x, y, touchAction );
            }
        }
        else if( this.touchMovementAllowed &&  CCControls.touchActionMoving( touchAction ) )
        {
            if( hitObject != this )
            {
                return this.handleProjectedTouch( hitObject, hitPosition, touch, touch_lost );
            }
            else
            {
                var relativeHitPosition = vec3.create();
                relativeHitPosition[0] = hitPosition[0] - position[0];
                relativeHitPosition[1] = hitPosition[1] - position[1];
                var x = relativeHitPosition[0] / collisionBounds[0];
                var y = relativeHitPosition[1] / collisionBounds[1];
                this.touchActionPressed( x, y, touchAction );
            }
        }
        else
        {   
            // Ensure we have a good touch release
            if( touchAction == touch_released )
            {
				if( this.touchMovementAllowed == false )
                {
					var absDeltaX = Math.abs( touch.totalDeltaX );
					var absDeltaY = Math.abs( touch.totalDeltaY );
					if( absDeltaX > CC_TOUCH_TO_MOVEMENT_THRESHOLD || absDeltaY > CC_TOUCH_TO_MOVEMENT_THRESHOLD )
					{
						return this.handleProjectedTouch( hitObject, hitPosition, touch, touch_lost );
					}
				}
                
                // If we have a good first touch, register it.
                var relativeHitPosition = vec3.create();
                relativeHitPosition[0] = hitPosition[0] - position[0];
                relativeHitPosition[1] = hitPosition[1] - position[1];
                var x = relativeHitPosition[0] / collisionBounds[0];
                var y = relativeHitPosition[1] / collisionBounds[1];
                this.touchActionPressed( x, y, touch_released );
            }
            
            this.touchActionRelease( touchAction );
            this.touching = false;
            
            if( touchAction == touch_released )
            {
                return 2;
            }
        }
    }
    
    return 0;
}

CCTile3DButton.prototype.touchActionPressed = function(x, y, touchAction)
{   
    // If we can press down on the tile
    if( this.touchDepressRange > 0.0 )
    {
        this.touchDepressInterpolator.pushV3( this.touchDepressPosition, vec3.create( [0.0, 0.0, -this.touchDepressDepth] ), true );
    }

    if( touchAction > touch_pressed && touchAction < touch_released )
    {
        this.onTouchMove();
    }
}

CCTile3DButton.prototype.touchActionRelease = function(touchAction)
{
    if( touchAction == touch_released )
    {
        this.touchReleased = true;
    }
    else 
    {
        this.handleTouchRelease();
		if( touchAction == touch_lost )
        {
            this.onTouchLoss();
        }
    }
}

CCTile3DButton.prototype.handleTouchRelease = function()
{
    // If the touch has been released successfully we fire our callback
    if( this.touchReleased )
    {
        this.touchReleased = false;
        this.onTouchRelease();
    }

	// Touch depress: On releasse
    if( this.touchDepressRange > 0.0 )
    {
        this.touchDepressInterpolator.pushV3( this.touchDepressPosition, vec3.create(), true );
    }
}


CCTile3DButton.prototype.allowTouchRotation = function(allow) 
{ 
	this.touchRotationAllowed = allow;
}


CCTile3DButton.prototype.allowTouchMovement = function(allow)
{
	this.touchMovementAllowed = allow;
}
    
	
CCTile3DButton.prototype.setTouchRotationSpeed = function(speed)
{
	this.touchRotationSpeed = speed;
}
    
	
CCTile3DButton.prototype.setTouchDepressRange = function(range)
{
	this.touchDepressRange = range;
}